xbmc/addons/Repository.cpp

```diff
@@ -52,9 +55,83 @@
 using namespace XFILE;
 using namespace ADDON;
 using namespace KODI::MESSAGING;
+using KODI::UTILITY::CDigest;
 
 using KODI::MESSAGING::HELPERS::DialogResponse;
 
+CRepository::ResolveResult CRepository::ResolvePathAndHash(const AddonPtr& addon) const
+{
+  std::string const& path = addon->Path();
+
+  auto dirIt = std::find_if(m_dirs.begin(), m_dirs.end(), [&path](DirInfo const& dir)
+  {
+    return URIUtils::PathHasParent(path, dir.datadir, true);
+  });
+  if (dirIt == m_dirs.end())
+  {
+    CLog::Log(LOGERROR, "Requested path {} not found in known repository directories", path);
+    return {};
+  }
+
+  if (!dirIt->hashes)
+  {
+    // We have a path, but need no hash
+    return {path, ""};
+  }
+
+  // Do not follow mirror redirect, we want the headers of the redirect response
+  CURL url{path};
+  url.SetProtocolOption("redirect-limit", "0");
+  CCurlFile file;
+  if (!file.Open(url))
+  {
+    CLog::Log(LOGERROR, "Could not fetch addon location and hash from {}", path);
+    return {};
+  }
+
+  // Return the location from the header so we don't have to look it up again
+  // (saves one request per addon install)
+  std::string location = file.GetRedirectURL();
+  // content-* headers are base64, convert to base16
+  std::string hash = StringUtils::ToHexadecimal(Base64::Decode(file.GetHttpHeader().GetValue("content-sha256")));
+
+  if (hash.empty())
+  {
+    // Expected hash, but none found -> fall back to old method
+    if (!FetchChecksum(path + ".sha256", hash))
+    {
+      CLog::Log(LOGERROR, "Failed to find hash for {} from HTTP header and in separate file", path);
+      return {};
```

Comment:
You can't just change this to fail if the .sha256 isn't there. That breaks all existing repos that has set 'hashes' to true.

Reply:
Also, hard-coding 'sha256' is not a good idea.. It might break tomorrow and you have the md5 situation again. I suggest putting the alg. in `<hashes>` for instance. That might solve the backwards compatibility issue as well, as you can just let 'true' mean md5.

Reply:
> That breaks all existing repos that has set 'hashes' to true.

How common is that? (out of curiosity)

> I suggest putting the alg. in <hashes> for instance. That might solve the backwards compatibility issue as well, as you can just let 'true' mean md5.

So the other repositories using MD5 hashes, what purpose are they using them for? Is there even any meaning in checking those hashes at all?

Reply:
Or is that what you meant with "MD5 verification can be dropped?" :-)